analyzer/evmtokenbalances/evm_token_balances.go

```diff
@@ -0,0 +1,299 @@
+package evmtokenbalances
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/iancoleman/strcase"
+	oasisConfig "github.com/oasisprotocol/oasis-sdk/client-sdk/go/config"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/oasisprotocol/oasis-indexer/analyzer"
+	"github.com/oasisprotocol/oasis-indexer/analyzer/modules"
+	"github.com/oasisprotocol/oasis-indexer/analyzer/util"
+	"github.com/oasisprotocol/oasis-indexer/config"
+	"github.com/oasisprotocol/oasis-indexer/log"
+	"github.com/oasisprotocol/oasis-indexer/storage"
+	"github.com/oasisprotocol/oasis-indexer/storage/oasis"
+)
+
+// Imagine a timeline starting from a `balanceOf` output `v0` followed by
+// a delta `d1` encountered in a `Transfer` log.
+// [v0][d1]
+//    ^   ^
+//    r0  r1
+// Consider two points in history:
+// - `r0` is the round that the token balances analyzer last downloaded the
+//   balance for.
+// - `r1` is a later round that the block scanner scanned where the balance
+//   changed.
+// Additionally, at this time:
+// - `last_download_round` is `r0`
+// - `last_mutate_round` is `r1`
+// - `balance` is `v0` + `d1`
+// - `extra` is some unimportant value, typically `d1`.
+//
+// Suppose at this point the token balances analyzer runs. It notices that
+// this account is _stale_, because `last_mutate_round` >
+// `last_download_round` on this account. It does the following atomically
+// (regardless of what the block scanner is doing):
+// 1. Set _next download round_ = `last_mutate_round` (= `r1`).
+// 2. Set `extra` = 0.
+// It then tries to download the balance (i.e. call balanceOf) at _next
+// download round_. This takes place while the rest of the indexer continues
+// to run.
+//
+// Suppose the block scanner then reaches round `r3` and encounters a
+// `Transfer` log.
+// [v0][d1][d2]
+//    ^   ^   ^
+//    r0  r1  r2
+// The block scanner does the following atomically (regardless of what the
+// token balances analyzer is doing):
+// 1. Set `last_mutate_round` = `r2`.
+// 2. Update `balance` = `balance` + `d2` (= `v0` + `d1` + `d2`).
+// 3. Update `extra` = `extra` + `d2` (= `d2`)
+//
+// Suppose the token balances analyzer then succeeds at downloading the
+// balance of the account, getting the balance `v1` (from round `r1`). It does
```

Comment:
This is a corner case race: while the balanceOf query was busy processing/querying at r1, r2 was produced, analyzed, and inserted into the DB.

Would it work to just drop the update in case this race occurs, and let the balanceOf updater query the contract again in the next iteration? That was the idea behind https://github.com/oasisprotocol/oasis-indexer/pull/284#discussion_r1093734302

Reply:
> would make the logic easier to reason about, IMO

don't wanna build a system with worse starvation behaviors. this design ensures forward progress even when an account is busy and it takes more than a block for the balance analyzer to make a "lap" through the stale balances queue.

your proposal and rationale is heard though

Reply:
oh and I forgot to add the updated timeline diagram here. adding

Reply:
diagram added :tipping_hand_person: